File: modules/03-paralleization/tutorial_langchain.py
import asyncio
import json
import os
from dataclasses import dataclass
from operator import itemgetter
from typing import Any, Final

from langchain_core.output_parsers import StrOutputParser
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.runnables import Runnable, RunnableParallel
from langchain_openai import ChatOpenAI

from tutorial_types import StepLogger, TutorialStepResult

STEP_ID: Final[str] = "step-2-langchain"
LIBRARY: Final[str] = "langchain"


@dataclass(frozen=True)
class LangchainConfig:
    api_key: str
    model_name: str
    base_url: str | None
    topic: str


def resolve_langchain_config() -> LangchainConfig | None:
    openai_api_key: str = os.getenv("OPENAI_API_KEY", "").strip()
    if openai_api_key:
        return LangchainConfig(
            api_key=openai_api_key,
            model_name="gpt-4o-mini",
            base_url=None,
            topic="Artificial Intelligence in Healthcare",
        )

    synthetic_api_key: str = os.getenv("SYNTHETIC_API_KEY", "").strip()
    if synthetic_api_key:
        return LangchainConfig(
            api_key=synthetic_api_key,
            model_name="hf:moonshotai/Kimi-K2.5",
            base_url="https://api.synthetic.new/openai/v1",
            topic="Artificial Intelligence in Healthcare",
        )

    return None


async def run_langchain_tutorial(log: StepLogger = print) -> TutorialStepResult:
    config: LangchainConfig | None = resolve_langchain_config()
    if config is None:
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="skipped",
            message="OPENAI_API_KEY or SYNTHETIC_API_KEY is missing.",
        )

    log("[langchain] STEP 1/3 setup")
    llm: ChatOpenAI = ChatOpenAI(
        model=config.model_name,
        temperature=0.2,
        api_key=config.api_key,
        base_url=config.base_url,
    )
    parser: StrOutputParser = StrOutputParser()

    summary_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Summarize the following topic in one sentence."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | parser
    )
    questions_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Generate two interesting questions for this topic."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | parser
    )
    keywords_chain: Runnable[Any, str] = (
        ChatPromptTemplate.from_messages(
            [
                ("system", "Extract three key terms as comma-separated text."),
                ("user", "{topic}"),
            ]
        )
        | llm
        | parser
    )

    parallel_chain: RunnableParallel = RunnableParallel(
        {
            "summary": summary_chain,
            "questions": questions_chain,
            "keywords": keywords_chain,
            "topic": itemgetter("topic"),
        }
    )

    log("[langchain] STEP 2/3 run")
    try:
        result: dict[str, str] = await parallel_chain.ainvoke({"topic": config.topic})
        output: str = json.dumps(result, indent=2, ensure_ascii=True)
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="completed",
            message="LangChain tutorial completed.",
            output=output,
        )
    except Exception as exception:  # noqa: BLE001
        return TutorialStepResult(
            step_id=STEP_ID,
            library=LIBRARY,
            status="failed",
            message=str(exception),
        )


async def _main() -> None:
    result: TutorialStepResult = await run_langchain_tutorial()
    print(json.dumps(result.__dict__, indent=2, ensure_ascii=True))


if __name__ == "__main__":
    asyncio.run(_main())

File: modules/03-paralleization/tutorial.langchain.ts
import { StringOutputParser } from "@langchain/core/output_parsers";
import { ChatPromptTemplate } from "@langchain/core/prompts";
import { RunnableMap } from "@langchain/core/runnables";
import { ChatOpenAI } from "@langchain/openai";
import { env } from "node:process";
import { string } from "zod/v4";

type ModelConfig = {
  apiKey: string;
  modelName: string;
  baseUrl: string;
};

function resolveModelConfig(): ModelConfig {
  return {
    apiKey: String(env.API_KEY),
    modelName: String(env.MODEL),
    baseUrl: String(env.BASE_URL),
  } as ModelConfig;
}

async function runLanchainExaple(): Promise<string> {
  const modelConfig: ModelConfig = resolveModelConfig();

  const model: ChatOpenAI = new ChatOpenAI({
    apiKey: modelConfig.apiKey,
    model: modelConfig.modelName,
    temperature: 0.2,
    configuration: { baseURL: modelConfig.baseUrl },
  });

  const parser: StringOutputParser = new StringOutputParser();

  const summaryPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Summarize the following topic in one sentence."],
      ["user", "{topic1}{topic2}"],
    ]);

  const questikonPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Generate two interesting questions for this topic."],
      ["user", "{topic1}{topic2}"],
    ]);

  const keywordsPrompt: ReturnType<typeof ChatPromptTemplate.fromMessages> =
    ChatPromptTemplate.fromMessages([
      ["system", "Extract three key terms as comma-separated text."],
      ["user", "{topic1}"],
    ]);

  /**
   * python의 RunnableParallel 과 동일한 영학을 한다
   * Runnable 객체를 병렬로 실행하고 각 결과 값을 key, value 객체로 묶어서 반환하는 핵심 컴포넌트임
   *
   * 단일 입력 데이터를 RunnableMap 내부에 정의된 모든 하위 Runnable 컴포넌트에 동일하게 전달함.
   * 주로 다중 LLM 호출이나 여러 검색 엔진 동시 조회와 가은 I/O 바운드 작업의 전체 지연 시간을 최소화 하는데 사용된다.
   *
   * 주요 특징 및 동작 방식:
   * 병렬 처리 (Parallel Execution): 내부에 선언된 모든 작업이 비동기적으로 동시에 실행됩니다.
   * 전체 실행 시간은 선언된 작업 중 가장 오래 걸리는 작업 소요 시간과 거의 동일하게 최적화 됩니다.
   * const parallelmMap = RunnableMap.from({
   * summary: summaryPrompt.pip(model).pipe(parser)},
   * question: questikonPrompt.pipe(model).pipe(parser),
   * keywords: keywordsPrompt.pipe(model).pipe(parser)
   * })
   * 이런 형식으로 나올대 데이터 출력 형식 고정해준다
   */

  const parallelMap: RunnableMap<
    {
      topic1: string;
      topic2: string;
    },
    {
      summary: string;
      questions: string;
      keywords: string;
    }
  > = RunnableMap.from({
    summary: summaryPrompt.pipe(model).pipe(parser),
    questions: questikonPrompt.pipe(model).pipe(parser),
    keywords: keywordsPrompt.pipe(model).pipe(parser),
  });

  const result: { summary: string; questions: string; keywords: string } =
    await parallelMap.invoke({
      topic1: "Play alone in my room. and writing some code in my room.",
      topic2: "so i selpt in my room",
    });
  const output: string = JSON.stringify(result, null, 2);

  return output;
}

console.log(await runLanchainExaple());
